phabricatoremails/settings.py

```diff
@@ -0,0 +1,122 @@
+import configparser
+import pathlib
+import smtplib
+from configparser import ConfigParser
+from dataclasses import dataclass
+from logging import Logger
+from typing import Any
+
+from phabricatoremails.db import Db
+from phabricatoremails.logging import create_dev_logger, create_logger
+from phabricatoremails.mail import SesMail, SmtpMail, FsMail
+from phabricatoremails.progress import StaticProgress, PhabricatorProgress
+from phabricatoremails.source import FileSource, PhabricatorSource
+from sqlalchemy import create_engine
+
+
+def _parse_logger(config: ConfigParser):
+    """Provides a logger set up according to our configuration
+
+    The regular logger is noisy and implements the MozLog JSON format, so we also
+    have a "dev" logger that simply prints out each log without any annotations.
+    The correct logger is chosen based on the provided configuration file.
+    """
+    if config.has_section("dev"):
+        return create_dev_logger()
+    else:
+        return create_logger()
+
+
+def _parse_source(config: ConfigParser, logger: Logger):
```

Comment:
This method mixes some dev and non-dev functionality without good separation (or validation). Those two should be a little more succinctly defined. For example, you are using a `dev.since_key` and `dev.limit` even after yuo return `dev_source_file`. 

Is this meant to only be used in a dev setting? If so, please make that more clear; otherwise the functionality should be more clearly separated.

- [ ] Add Args and Returns entries to docstring
- [ ] Rename method to be more descriptive (it is returning "source" and "progress", and does not take "source" so the name doesn't make sense)

Reply:
The solution here is a bit of a :thinking:.

Here's the criteria in which each implementation is used:
(Note: `PhabricatorProgress` is now `PhabricatorWorker` and `StaticProgress` is now `RunOnceWorker`).
_ | `FileSource` | `PhabricatorSource`
--|--|--
**PhabricatorWorker** | _not possible_ | default
**RunOnceWorker** | `dev.file` is set | `dev.since_key` is set (and `dev.file` isn't)

Importantly, it's not possible to have `FileSource` and `PhabricatorWorker`. I considered splitting up `_parse_source()` into `_parse_source()` and `_parse_worker()`, but then it would hide the constraints on the possible combinations.

Additionally, you're right that both dev and non-dev functionality is being determined here. It's ok that we are using `dev.since_key` after we're done with `dev.source_file`, because they're defining different development environment configurations (hopefully this is more clear with the table above^).

There's room for improvement here, I'll chew on it.